### DISCUSSION TITLE
What's the problem with this code (hyperbolic Delaunay)?

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#6593] [Q&A]
Hello,
 
 [Here is my code](https://gist.github.com/stla/9e49a91929aeef7d34a001ae666ec021). I call it from R and it works fine. However, when it is checked with Valgrind (a tool to check memory problems in R packages), there are some issues, such as:

```
==539271== Conditional jump or move depends on uninitialised value(s)
==539271==    at 0x5785AC92: operator() (R-devel/site-library/RcppCGAL/include/CGAL/Hyperbolic_triangulation_2/internal/Hyperbolic_Delaunay_triangulation_traits_2_functions.h:180)
==539271==    by 0x5785AC92: mark_face (R-devel/site-library/RcppCGAL/include/CGAL/Hyperbolic_Delaunay_triangulation_2.h:802)
==539271==    by 0x5785AC92: CGAL::Hyperbolic_Delaunay_triangulation_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_with_id_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_vertex_base_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_ds_vertex_base_2<void> > >, CGAL::Hyperbolic_triangulation_face_base_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_ds_face_base_2<void> > > >::mark_finite_non_hyperbolic_faces() const (R-devel/site-library/RcppCGAL/include/CGAL/Hyperbolic_Delaunay_triangulation_2.h:734)
==539271==    by 0x578644CD: insert<__gnu_cxx::__normal_iterator<CGAL::Point_2<CGAL::Epick>*, std::vector<CGAL::Point_2<CGAL::Epick>, std::allocator<CGAL::Point_2<CGAL::Epick> > > > > (R-devel/site-library/RcppCGAL/include/CGAL/Hyperbolic_Delaunay_triangulation_2.h:450)
==539271==    by 0x578644CD: Rcpp::Vector<19, Rcpp::PreserveStorage> hdelaunay_cpp<CGAL::Hyperbolic_Delaunay_triangulation_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_data_structure_2<CGAL::Triangulation_vertex_base_with_id_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_vertex_base_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_ds_vertex_base_2<void> > >, CGAL::Hyperbolic_triangulation_face_base_2<CGAL::Hyperbolic_Delaunay_triangulation_traits_2<CGAL::Epick>, CGAL::Triangulation_ds_face_base_2<void> > > >, CGAL::Point_2<CGAL::Epick> >(Rcpp::Matrix<14, Rcpp::PreserveStorage>) (packages/tests-vg/gyro/src/hdelaunay.cpp:60)
==539271==    by 0x57830422: hdelaunay_K(Rcpp::Matrix<14, Rcpp::PreserveStorage>) (packages/tests-vg/gyro/src/hdelaunay.cpp:118)
```

Because of this problem I can't publish my package :-(. Do you have an idea?

**Comment by stla** [comment]
I found [another way to proceed](https://gist.github.com/stla/9e49a91929aeef7d34a001ae666ec021#file-hdelaunay2-cpp). Hopefully this will work now.

**Comment by stla** [comment]
Does not work. Valgrind complains at the line calling `insert`:

`hdt.insert(hpts.begin(), hpts.end());`

I wanted to try `reserve(hdt, nvertices, nfaces, nedges)` but I don't understand how to call this function. What is its namespace?

**Reply by afabri** [reply]
What do you refer to with  `reserve()` ?

**Reply by stla** [reply]
Hello,
I'm not sure to understand your question. Valgrind seems to complain because my hyperbolic Delaunay triangulation (`hdt`) object is not initialized. So I'm trying to use `reserve` which is defined for the `MutableFaceGraph`s.

**Reply by afabri** [reply]
Triangulations are not mutable. At least that is not stated [here](https://doc.cgal.org/latest/BGL/group__PkgBGLTraits.html).

**Reply by stla** [reply]
Ah ok. How could I allocate memory? Can I use **malloc**? Everything works fine, that's just Valgrind who complains.

**Comment by stla** [comment]
It is claimed in the doc that `push_back` is equivalent to `insert`. However my code compiles with `insert` but not with `push_back` (error: Vertex_handle inaccessible within this context).

**Reply by afabri** [reply]
You write that because it should be inherited from the base class `Delaunay_triangulation_2`?

**Reply by stla** [reply]
Yes, I'm refering to [this point](https://doc.cgal.org/latest/Triangulation_2/classCGAL_1_1Delaunay__triangulation__2.html#a392d4df302f4f255eada59588c84b86e) in the doc.

**Reply by afabri** [reply]
It would be easy to add. It is also well possible that other functions of the base class are missing.

**Comment by afabri** [comment]
When I compile the example with the address sanitizer of VC++ enabled I also get error messages.  So we should maybe have a closer look.

**Reply by afabri** [reply]
Well, I do not have it with clang and its address sanitizer.

**Comment by stla** [comment]
I've done a minimal example:

```cpp
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Triangulation_data_structure_2.h>
#include <CGAL/Hyperbolic_Delaunay_triangulation_2.h>
#include <CGAL/Hyperbolic_Delaunay_triangulation_traits_2.h>
#include <CGAL/Triangulation_vertex_base_with_id_2.h>

typedef CGAL::Exact_predicates_inexact_constructions_kernel K;
typedef CGAL::Hyperbolic_Delaunay_triangulation_traits_2<K> HDtt;
typedef HDtt::Point_2 HPoint;
typedef CGAL::Triangulation_data_structure_2<
  CGAL::Triangulation_vertex_base_with_id_2<HDtt>,
  CGAL::Hyperbolic_triangulation_face_base_2<HDtt>>
    HTds;
typedef CGAL::Hyperbolic_Delaunay_triangulation_2<HDtt, HTds> HDt;

int main(){
  HDt hdt;
  hdt.insert(HPoint(0.1, 0.1));
  hdt.insert(HPoint(0.8, 0.2));
  hdt.insert(HPoint(0.4, 0.6));
  hdt.insert(HPoint(-0.5, -0.5));
  return 0;
} 
```

[Here are](https://gist.github.com/stla/9e49a91929aeef7d34a001ae666ec021#file-hdelaunay_valgrind-txt) the Valgrind complains. Functions whose name start with `R` are from R. 

So it really seems the problem is not from my code (this time), this is from CGAL.

I said something wrong here:

> Valgrind (a tool to check memory problems in R packages)

Valgrind has nothing to do with R. This is a tool to check C/C++ code. There's also UBSAN or ASAN which are similar tools, that you can use via Docker.

**Comment by stla** [comment]
I checked the Euclidean Delaunay triangulation with Valgrind. No issue, clean.
